editor/src/components/editor/store/collaborative-editing.ts

```diff
@@ -0,0 +1,363 @@
+  updatesAt: Array<number>
+  deleteFrom: number | null
+}
+
+function updateEntireProjectContents(
+  changeEvent: Y.YMapEvent<any>,
+  dispatch: EditorDispatch,
+): void {
+  let actions: Array<EditorAction> = []
+  const entriesIterator = changeEvent.keys.entries()
+  let entriesIteratorResult = entriesIterator.next()
+  while (!entriesIteratorResult.done) {
+    const [filename, changeEntry] = entriesIteratorResult.value
+    switch (changeEntry.action) {
+      case 'update':
+      case 'add':
+        if (changeEntry.newValue != null) {
+          actions.push(
+            updateTopLevelElements(
+              filename,
+              (changeEntry.newValue as CollabTextFileTopLevelElements).toArray(),
+            ),
+          )
+        }
+        break
+      case 'delete':
+        actions.push(deleteFile(filename))
+        break
+      default:
+        throw new Error(`Unhandled change entry action: ${changeEntry.action}`)
+    }
+    entriesIteratorResult = entriesIterator.next()
+  }
+  dispatch(actions)
+}
+
+function updateTopLevelElementsOfFile(
+  session: CollaborativeEditingSupportSession,
+  filePath: string,
+  changeEvent: Y.YEvent<any>,
+  dispatch: EditorDispatch,
+): void {
+  const file = session.projectContents.get(filePath)
+  if (file != null) {
+    const oldTopLevelElements = file.get('topLevelElements') as CollabTextFileTopLevelElements
+    let newTopLevelElements: Array<TopLevelElement> = []
+    let readIndex = 0
+    for (const delta of changeEvent.delta) {
+      if (delta.retain != undefined) {
+        const elementsToPush = oldTopLevelElements.slice(readIndex, readIndex + delta.retain)
+        newTopLevelElements.push(...elementsToPush)
+        readIndex += delta.retain
+      }
+      if (delta.insert != null && Array.isArray(delta.insert)) {
+        newTopLevelElements.push(...delta.insert)
+      }
+      if (delta.delete != undefined) {
+        readIndex += delta.delete
+      }
+    }
+
+    if (readIndex < oldTopLevelElements.length) {
+      // There is an implicit retain for the remainder of the items
+      const elementsToPush = oldTopLevelElements.slice(readIndex)
+      newTopLevelElements.push(...elementsToPush)
+    }
+
+    dispatch([updateTopLevelElements(filePath, newTopLevelElements)])
+  }
+}
+
+function arrayChanges(updatesAt: Array<number>, deleteFrom: number | null): ArrayChanges {
+  return {
+    updatesAt: updatesAt,
+    deleteFrom: deleteFrom,
+  }
+}
+
+export function calculateArrayChanges<T>(
+  from: Array<T>,
+  into: Y.Array<T>,
+  equals: (from: T, into: T) => boolean = (fromToCheck, intoToCheck) => fromToCheck === intoToCheck,
+): ArrayChanges {
+  let updatesAt: Array<number> = []
+  for (let index: number = 0; index < Math.max(from.length, into.length); index++) {
+    if (!equals(from[index], into.get(index))) {
+      updatesAt.push(index)
+    }
+  }
+
+  let deleteFrom: number | null = null
+  if (into.length > from.length) {
+    deleteFrom = into.length
```

Comment:
Shouldn't this be `deleteFrom = from.length`?

Reply:
Fixed in 8dd00397d.